File: cpp_engine/src/pattern_recognition.cpp
/**
 * Pattern Recognition Implementation
 * High-performance candlestick pattern detection
 */

#include "pattern_recognition.hpp"
#include <cmath>
#include <algorithm>

namespace trading_engine {

PatternRecognition::PatternRecognition(double body_min_ratio, double wick_max_ratio)
    : body_min_ratio_(body_min_ratio), wick_max_ratio_(wick_max_ratio) {}

// Helper functions
double PatternRecognition::body_size(const Bar& bar) const {
    return std::abs(bar.close - bar.open);
}

double PatternRecognition::upper_wick(const Bar& bar) const {
    return bar.high - std::max(bar.open, bar.close);
}

double PatternRecognition::lower_wick(const Bar& bar) const {
    return std::min(bar.open, bar.close) - bar.low;
}

double PatternRecognition::total_range(const Bar& bar) const {
    return bar.high - bar.low;
}

bool PatternRecognition::is_bullish(const Bar& bar) const {
    return bar.close > bar.open;
}

bool PatternRecognition::is_bearish(const Bar& bar) const {
    return bar.close < bar.open;
}

// Doji pattern - small body, long wicks
bool PatternRecognition::is_doji(const Bar& bar) const {
    double range = total_range(bar);
    if (range < 1e-10) return false;
    
    double body = body_size(bar);
    double body_ratio = body / range;
    
    // Doji has very small body (< 10% of range)
    return body_ratio < 0.1;
}

// Hammer - small body at top, long lower wick (bullish reversal)
bool PatternRecognition::is_hammer(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 5 || idx >= bars.size()) return false;
    
    const Bar& bar = bars[idx];
    double range = total_range(bar);
    if (range < 1e-10) return false;
    
    double body = body_size(bar);
    double lower = lower_wick(bar);
    double upper = upper_wick(bar);
    
    // Check hammer conditions
    bool small_body = body < range * 0.35;
    bool long_lower_wick = lower > body * 2.0;
    bool small_upper_wick = upper < body * 0.5;
    
    if (!(small_body && long_lower_wick && small_upper_wick)) return false;
    
    // Verify downtrend (previous 5 bars)
    double trend_sum = 0;
    for (size_t i = idx - 5; i < idx; ++i) {
        trend_sum += bars[i].close - bars[i].open;
    }
    
    return trend_sum < 0;  // Should be in a downtrend
}

// Shooting Star - small body at bottom, long upper wick (bearish reversal)
bool PatternRecognition::is_shooting_star(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 5 || idx >= bars.size()) return false;
    
    const Bar& bar = bars[idx];
    double range = total_range(bar);
    if (range < 1e-10) return false;
    
    double body = body_size(bar);
    double lower = lower_wick(bar);
    double upper = upper_wick(bar);
    
    // Check shooting star conditions
    bool small_body = body < range * 0.35;
    bool long_upper_wick = upper > body * 2.0;
    bool small_lower_wick = lower < body * 0.5;
    
    if (!(small_body && long_upper_wick && small_lower_wick)) return false;
    
    // Verify uptrend (previous 5 bars)
    double trend_sum = 0;
    for (size_t i = idx - 5; i < idx; ++i) {
        trend_sum += bars[i].close - bars[i].open;
    }
    
    return trend_sum > 0;  // Should be in an uptrend
}

// Engulfing pattern
int PatternRecognition::is_engulfing(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 1 || idx >= bars.size()) return 0;
    
    const Bar& curr = bars[idx];
    const Bar& prev = bars[idx - 1];
    
    // Bullish engulfing: prev bearish, curr bullish, curr body engulfs prev
    if (is_bearish(prev) && is_bullish(curr)) {
        if (curr.open < prev.close && curr.close > prev.open) {
            return 1;  // Bullish
        }
    }
    
    // Bearish engulfing: prev bullish, curr bearish, curr body engulfs prev
    if (is_bullish(prev) && is_bearish(curr)) {
        if (curr.open > prev.close && curr.close < prev.open) {
            return -1;  // Bearish
        }
    }
    
    return 0;
}

// Three White Soldiers - three consecutive bullish candles
bool PatternRecognition::is_three_soldiers(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 2 || idx >= bars.size()) return false;
    
    const Bar& c1 = bars[idx - 2];
    const Bar& c2 = bars[idx - 1];
    const Bar& c3 = bars[idx];
    
    // All three must be bullish with strong bodies
    if (!is_bullish(c1) || !is_bullish(c2) || !is_bullish(c3)) return false;
    
    // Each close higher than previous
    if (c2.close <= c1.close || c3.close <= c2.close) return false;
    
    // Each open within previous body
    if (c2.open < c1.open || c2.open > c1.close) return false;
    if (c3.open < c2.open || c3.open > c2.close) return false;
    
    // Small upper wicks
    double r1 = total_range(c1), r2 = total_range(c2), r3 = total_range(c3);
    if (r1 > 0 && upper_wick(c1) / r1 > 0.3) return false;
    if (r2 > 0 && upper_wick(c2) / r2 > 0.3) return false;
    if (r3 > 0 && upper_wick(c3) / r3 > 0.3) return false;
    
    return true;
}

// Three Black Crows - three consecutive bearish candles
bool PatternRecognition::is_three_crows(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 2 || idx >= bars.size()) return false;
    
    const Bar& c1 = bars[idx - 2];
    const Bar& c2 = bars[idx - 1];
    const Bar& c3 = bars[idx];
    
    // All three must be bearish
    if (!is_bearish(c1) || !is_bearish(c2) || !is_bearish(c3)) return false;
    
    // Each close lower than previous
    if (c2.close >= c1.close || c3.close >= c2.close) return false;
    
    // Each open within previous body
    if (c2.open > c1.open || c2.open < c1.close) return false;
    if (c3.open > c2.open || c3.open < c2.close) return false;
    
    // Small lower wicks
    double r1 = total_range(c1), r2 = total_range(c2), r3 = total_range(c3);
    if (r1 > 0 && lower_wick(c1) / r1 > 0.3) return false;
    if (r2 > 0 && lower_wick(c2) / r2 > 0.3) return false;
    if (r3 > 0 && lower_wick(c3) / r3 > 0.3) return false;
    
    return true;
}

// Morning Star - bullish reversal (3 candle pattern)
bool PatternRecognition::is_morning_star(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 2 || idx >= bars.size()) return false;
    
    const Bar& c1 = bars[idx - 2];  // First candle - bearish
    const Bar& c2 = bars[idx - 1];  // Second candle - small body (star)
    const Bar& c3 = bars[idx];      // Third candle - bullish
    
    // First must be bearish, third must be bullish
    if (!is_bearish(c1) || !is_bullish(c3)) return false;
    
    // Middle candle should have small body
    double r2 = total_range(c2);
    if (r2 > 0 && body_size(c2) / r2 > 0.4) return false;
    
    // Gap down from first to second
    if (c2.high > c1.low) return false;
    
    // Third candle closes above midpoint of first
    double c1_mid = (c1.open + c1.close) / 2;
    if (c3.close < c1_mid) return false;
    
    return true;
}

// Evening Star - bearish reversal (3 candle pattern)
bool PatternRecognition::is_evening_star(const std::vector<Bar>& bars, size_t idx) const {
    if (idx < 2 || idx >= bars.size()) return false;
    
    const Bar& c1 = bars[idx - 2];  // First candle - bullish
    const Bar& c2 = bars[idx - 1];  // Second candle - small body (star)
    const Bar& c3 = bars[idx];      // Third candle - bearish
    
    // First must be bullish, third must be bearish
    if (!is_bullish(c1) || !is_bearish(c3)) return false;
    
    // Middle candle should have small body
    double r2 = total_range(c2);
    if (r2 > 0 && body_size(c2) / r2 > 0.4) return false;
    
    // Gap up from first to second
    if (c2.low < c1.high) return false;
    
    // Third candle closes below midpoint of first
    double c1_mid = (c1.open + c1.close) / 2;
    if (c3.close > c1_mid) return false;
    
    return true;
}

// Batch detection
std::vector<PatternResult> PatternRecognition::detect_all(
    const std::vector<double>& open,
    const std::vector<double>& high,
    const std::vector<double>& low,
    const std::vector<double>& close,
    size_t lookback
) const {
    std::vector<PatternResult> results;
    
    size_t n = std::min({open.size(), high.size(), low.size(), close.size()});
    if (n < 3) return results;
    
    // Convert to Bar format
    std::vector<Bar> bars(n);
    for (size_t i = 0; i < n; ++i) {
        bars[i] = {open[i], high[i], low[i], close[i], 0.0, 0};
    }
    
    // Start from lookback
    size_t start = (n > lookback) ? n - lookback : 0;
    
    for (size_t i = start; i < n; ++i) {
        // Doji
        if (is_doji(bars[i])) {
            results.push_back({"Doji", "neutral", 0.7, static_cast<int>(i), 
                              "Indecision, potential reversal"});
        }
        
        // Hammer
        if (is_hammer(bars, i)) {
            results.push_back({"Hammer", "bullish", 0.75, static_cast<int>(i),
                              "Bullish reversal after downtrend"});
        }
        
        // Shooting Star
        if (is_shooting_star(bars, i)) {
            results.push_back({"Shooting Star", "bearish", 0.75, static_cast<int>(i),
                              "Bearish reversal after uptrend"});
        }
        
        // Engulfing
        int engulf = is_engulfing(bars, i);
        if (engulf == 1) {
            results.push_back({"Bullish Engulfing", "bullish", 0.8, static_cast<int>(i),
                              "Strong bullish reversal signal"});
        } else if (engulf == -1) {
            results.push_back({"Bearish Engulfing", "bearish", 0.8, static_cast<int>(i),
                              "Strong bearish reversal signal"});
        }
        
        // Three Soldiers
        if (is_three_soldiers(bars, i)) {
            results.push_back({"Three White Soldiers", "bullish", 0.85, static_cast<int>(i),
                              "Strong bullish continuation"});
        }
        
        // Three Crows
        if (is_three_crows(bars, i)) {
            results.push_back({"Three Black Crows", "bearish", 0.85, static_cast<int>(i),
                              "Strong bearish continuation"});
        }
        
        // Morning Star
        if (is_morning_star(bars, i)) {
            results.push_back({"Morning Star", "bullish", 0.85, static_cast<int>(i),
                              "Major bullish reversal"});
        }
        
        // Evening Star
        if (is_evening_star(bars, i)) {
            results.push_back({"Evening Star", "bearish", 0.85, static_cast<int>(i),
                              "Major bearish reversal"});
        }
    }
    
    return results;
}

// Aggregate signals
std::pair<double, double> PatternRecognition::aggregate_signals(
    const std::vector<PatternResult>& patterns,
    double min_confidence
) const {
    if (patterns.empty()) {
        return {0.0, 0.0};
    }
    
    double bullish_score = 0.0;
    double bearish_score = 0.0;
    int bullish_count = 0;
    int bearish_count = 0;
    
    for (const auto& p : patterns) {
        if (p.confidence < min_confidence) continue;
        
        if (p.type == "bullish") {
            bullish_score += p.confidence;
            ++bullish_count;
        } else if (p.type == "bearish") {
            bearish_score += p.confidence;
            ++bearish_count;
        }
    }
    
    double total = bullish_score + bearish_score;
    if (total < 1e-10) {
        return {0.0, 0.0};
    }
    
    // Signal: -1 to +1
    double signal = (bullish_score - bearish_score) / total;
    
    // Confidence based on count and score
    int total_count = bullish_count + bearish_count;
    double confidence = std::min(1.0, total / (total_count * 1.0));
    
    return {signal, confidence};
}

} // namespace trading_engine

File: indicators/pattern_recognition.py
# indicators/pattern_recognition.py
"""
Comprehensive Chart Pattern Recognition
Detects both candlestick patterns and classical chart patterns
"""

import pandas as pd
import numpy as np
from typing import List, Dict, Tuple


class CandlestickPatternDetector:
    """Detects candlestick patterns using price action rules"""
    
    def __init__(self, body_min_ratio=0.6, wick_max_ratio=0.3):
        self.body_min_ratio = body_min_ratio
        self.wick_max_ratio = wick_max_ratio
    
    def detect_doji(self, row) -> Tuple[bool, float]:
        """Detect Doji pattern - small body, long wicks"""
        body = abs(row['close'] - row['open'])
        total_range = row['high'] - row['low']
        
        if total_range == 0:
            return False, 0.0
        
        body_ratio = body / total_range
        
        # Doji: body is less than 10% of total range
        is_doji = body_ratio < 0.1
        confidence = 1.0 - (body_ratio / 0.1) if is_doji else 0.0
        
        return is_doji, min(confidence, 1.0)
    
    def detect_hammer(self, df, idx) -> Tuple[bool, float]:
        """Detect Hammer - small body at top, long lower wick"""
        if idx < 1:
            return False, 0.0
        
        row = df.iloc[idx]
        body = abs(row['close'] - row['open'])
        total_range = row['high'] - row['low']
        lower_wick = min(row['open'], row['close']) - row['low']
        upper_wick = row['high'] - max(row['open'], row['close'])
        
        if total_range == 0:
            return False, 0.0
        
        # Hammer criteria:
        # 1. Lower wick is at least 2x the body
        # 2. Upper wick is very small
        # 3. Appears in downtrend
        lower_wick_ratio = lower_wick / total_range
        body_ratio = body / total_range
        upper_wick_ratio = upper_wick / total_range
        
        is_hammer = (
            lower_wick >= body * 2 and
            upper_wick_ratio < 0.1 and
            body_ratio < 0.3 and
            lower_wick_ratio > 0.6
        )
        
        # Check if in downtrend (price declining over last 5 candles)
        if is_hammer and idx >= 5:
            trend = df.iloc[idx-5:idx]['close'].mean() > row['close']
            if not trend:
                is_hammer = False
        
        confidence = lower_wick_ratio * 1.2 if is_hammer else 0.0
        return is_hammer, min(confidence, 1.0)
    
    def detect_shooting_star(self, df, idx) -> Tuple[bool, float]:
        """Detect Shooting Star - small body at bottom, long upper wick"""
        if idx < 1:
            return False, 0.0
        
        row = df.iloc[idx]
        body = abs(row['close'] - row['open'])
        total_range = row['high'] - row['low']
        lower_wick = min(row['open'], row['close']) - row['low']
        upper_wick = row['high'] - max(row['open'], row['close'])
        
        if total_range == 0:
            return False, 0.0
        
        upper_wick_ratio = upper_wick / total_range
        body_ratio = body / total_range
        lower_wick_ratio = lower_wick / total_range
        
        is_shooting_star = (
            upper_wick >= body * 2 and
            lower_wick_ratio < 0.1 and
            body_ratio < 0.3 and
            upper_wick_ratio > 0.6
        )
        
        # Check if in uptrend
        if is_shooting_star and idx >= 5:
            trend = df.iloc[idx-5:idx]['close'].mean() < row['close']
            if not trend:
                is_shooting_star = False
        
        confidence = upper_wick_ratio * 1.2 if is_shooting_star else 0.0
        return is_shooting_star, min(confidence, 1.0)
    
    def detect_engulfing(self, df, idx) -> Tuple[str, float]:
        """Detect Bullish/Bearish Engulfing pattern"""
        if idx < 1:
            return None, 0.0
        
        prev = df.iloc[idx-1]
        curr = df.iloc[idx]
        
        prev_body = abs(prev['close'] - prev['open'])
        curr_body = abs(curr['close'] - curr['open'])
        
        # Bullish Engulfing
        if (prev['close'] < prev['open'] and  # Previous was bearish
            curr['close'] > curr['open'] and  # Current is bullish
            curr['open'] < prev['close'] and  # Opens below previous close
            curr['close'] > prev['open']):    # Closes above previous open
            
            engulfing_ratio = curr_body / prev_body if prev_body > 0 else 0
            confidence = min(engulfing_ratio / 2, 1.0)
            return 'bullish_engulfing', confidence
        
        # Bearish Engulfing
        if (prev['close'] > prev['open'] and  # Previous was bullish
            curr['close'] < curr['open'] and  # Current is bearish
            curr['open'] > prev['close'] and  # Opens above previous close
            curr['close'] < prev['open']):    # Closes below previous open
            
            engulfing_ratio = curr_body / prev_body if prev_body > 0 else 0
            confidence = min(engulfing_ratio / 2, 1.0)
            return 'bearish_engulfing', confidence
        
        return None, 0.0
    
    def detect_three_soldiers(self, df, idx) -> Tuple[bool, float]:
        """Detect Three White Soldiers - three consecutive bullish candles"""
        if idx < 2:
            return False, 0.0
        
        candles = df.iloc[idx-2:idx+1]
        
        # All three must be bullish
        all_bullish = all(candles['close'] > candles['open'])
        
        # Each closes higher than previous
        ascending = (candles.iloc[1]['close'] > candles.iloc[0]['close'] and
                    candles.iloc[2]['close'] > candles.iloc[1]['close'])
        
        # Each opens within previous candle's body
        opens_within_body = (
            candles.iloc[1]['open'] >= candles.iloc[0]['open'] and
            candles.iloc[1]['open'] <= candles.iloc[0]['close'] and
            candles.iloc[2]['open'] >= candles.iloc[1]['open'] and
            candles.iloc[2]['open'] <= candles.iloc[1]['close']
        )
        
        is_three_soldiers = all_bullish and ascending and opens_within_body
        
        # Confidence based on candle sizes
        if is_three_soldiers:
            avg_body = candles.apply(lambda r: abs(r['close'] - r['open']), axis=1).mean()
            avg_range = (candles['high'] - candles['low']).mean()
            confidence = avg_body / avg_range if avg_range > 0 else 0
        else:
            confidence = 0.0
        
        return is_three_soldiers, min(confidence, 1.0)
    
    def detect_three_crows(self, df, idx) -> Tuple[bool, float]:
        """Detect Three Black Crows - three consecutive bearish candles"""
        if idx < 2:
            return False, 0.0
        
        candles = df.iloc[idx-2:idx+1]
        
        # All three must be bearish
        all_bearish = all(candles['close'] < candles['open'])
        
        # Each closes lower than previous
        descending = (candles.iloc[1]['close'] < candles.iloc[0]['close'] and
                     candles.iloc[2]['close'] < candles.iloc[1]['close'])
        
        # Each opens within previous candle's body
        opens_within_body = (
            candles.iloc[1]['open'] <= candles.iloc[0]['open'] and
            candles.iloc[1]['open'] >= candles.iloc[0]['close'] and
            candles.iloc[2]['open'] <= candles.iloc[1]['open'] and
            candles.iloc[2]['open'] >= candles.iloc[1]['close']
        )
        
        is_three_crows = all_bearish and descending and opens_within_body
        
        if is_three_crows:
            avg_body = candles.apply(lambda r: abs(r['close'] - r['open']), axis=1).mean()
            avg_range = (candles['high'] - candles['low']).mean()
            confidence = avg_body / avg_range if avg_range > 0 else 0
        else:
            confidence = 0.0
        
        return is_three_crows, min(confidence, 1.0)
    
    def detect_morning_star(self, df, idx) -> Tuple[bool, float]:
        """Detect Morning Star - bullish reversal pattern"""
        if idx < 2:
            return False, 0.0
        
        first = df.iloc[idx-2]
        second = df.iloc[idx-1]
        third = df.iloc[idx]
        
        # First candle: Large bearish
        first_bearish = first['close'] < first['open']
        first_body = abs(first['close'] - first['open'])
        
        # Second candle: Small body (star)
        second_body = abs(second['close'] - second['open'])
        second_is_small = second_body < first_body * 0.3
        
        # Third candle: Large bullish
        third_bullish = third['close'] > third['open']
        third_body = abs(third['close'] - third['open'])
        third_closes_high = third['close'] > (first['open'] + first['close']) / 2
        
        is_morning_star = (
            first_bearish and
            second_is_small and
            third_bullish and
            third_closes_high
        )
        
        if is_morning_star:
            confidence = min((third_body / first_body) * 0.7, 1.0)
        else:
            confidence = 0.0
        
        return is_morning_star, confidence
    
    def detect_evening_star(self, df, idx) -> Tuple[bool, float]:
        """Detect Evening Star - bearish reversal pattern"""
        if idx < 2:
            return False, 0.0
        
        first = df.iloc[idx-2]
        second = df.iloc[idx-1]
        third = df.iloc[idx]
        
        # First candle: Large bullish
        first_bullish = first['close'] > first['open']
        first_body = abs(first['close'] - first['open'])
        
        # Second candle: Small body (star)
        second_body = abs(second['close'] - second['open'])
        second_is_small = second_body < first_body * 0.3
        
        # Third candle: Large bearish
        third_bearish = third['close'] < third['open']
        third_body = abs(third['close'] - third['open'])
        third_closes_low = third['close'] < (first['open'] + first['close']) / 2
        
        is_evening_star = (
            first_bullish and
            second_is_small and
            third_bearish and
            third_closes_low
        )
        
        if is_evening_star:
            confidence = min((third_body / first_body) * 0.7, 1.0)
        else:
            confidence = 0.0
        
        return is_evening_star, confidence


def detect_candlestick_patterns(df: pd.DataFrame, lookback=100) -> List[Dict]:
    """
    Detect all candlestick patterns in the dataframe
    
    Returns:
        List of detected patterns with metadata
    """
    detector = CandlestickPatternDetector()
    patterns = []
    
    # Only check recent candles to save time
    start_idx = max(0, len(df) - lookback)
    
    for idx in range(start_idx, len(df)):
        row = df.iloc[idx]
        
        # Doji
        is_doji, conf = detector.detect_doji(row)
        if is_doji and conf > 0.6:
            patterns.append({
                'pattern': 'doji',
                'type': 'candlestick',
                'direction': 0,  # Neutral
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Hammer
        is_hammer, conf = detector.detect_hammer(df, idx)
        if is_hammer and conf > 0.6:
            patterns.append({
                'pattern': 'hammer',
                'type': 'candlestick',
                'direction': 1,  # Bullish
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Shooting Star
        is_shooting, conf = detector.detect_shooting_star(df, idx)
        if is_shooting and conf > 0.6:
            patterns.append({
                'pattern': 'shooting_star',
                'type': 'candlestick',
                'direction': -1,  # Bearish
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Engulfing
        eng_type, conf = detector.detect_engulfing(df, idx)
        if eng_type and conf > 0.6:
            patterns.append({
                'pattern': eng_type,
                'type': 'candlestick',
                'direction': 1 if 'bullish' in eng_type else -1,
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Three Soldiers
        is_soldiers, conf = detector.detect_three_soldiers(df, idx)
        if is_soldiers and conf > 0.6:
            patterns.append({
                'pattern': 'three_white_soldiers',
                'type': 'candlestick',
                'direction': 1,
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Three Crows
        is_crows, conf = detector.detect_three_crows(df, idx)
        if is_crows and conf > 0.6:
            patterns.append({
                'pattern': 'three_black_crows',
                'type': 'candlestick',
                'direction': -1,
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Morning Star
        is_morning, conf = detector.detect_morning_star(df, idx)
        if is_morning and conf > 0.6:
            patterns.append({
                'pattern': 'morning_star',
                'type': 'candlestick',
                'direction': 1,
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
        
        # Evening Star
        is_evening, conf = detector.detect_evening_star(df, idx)
        if is_evening and conf > 0.6:
            patterns.append({
                'pattern': 'evening_star',
                'type': 'candlestick',
                'direction': -1,
                'confidence': conf,
                'index': idx,
                'timestamp': row.get('timestamp', idx),
                'price': row['close']
            })
    
    return patterns


def detect_chart_patterns(df: pd.DataFrame, lookback=100) -> List[Dict]:
    """
    Detect classical chart patterns (triangles, H&S, double tops, etc.)
    """
    from .chart_patterns import detect_chart_patterns as detect_classical_patterns
    return detect_classical_patterns(df, lookback)


def detect_all_patterns(df: pd.DataFrame, lookback=100) -> List[Dict]:
    """
    Detect all patterns (candlestick + chart patterns)
    
    Args:
        df: DataFrame with OHLC data
        lookback: Number of candles to analyze
    
    Returns:
        List of all detected patterns
    """
    candlestick_patterns = detect_candlestick_patterns(df, lookback)
    chart_patterns = detect_chart_patterns(df, lookback)
    
    all_patterns = candlestick_patterns + chart_patterns
    
    # Sort by index (most recent first)
    all_patterns.sort(key=lambda x: x['index'], reverse=True)
    
    return all_patterns


def aggregate_pattern_signals(patterns: List[Dict], min_confidence=0.7) -> Dict:
    """
    Aggregate multiple pattern signals into a single trading signal
    
    Returns:
        {
            'signal': 1 (buy), -1 (sell), or 0 (neutral),
            'confidence': overall confidence,
            'pattern_count': number of patterns detected,
            'details': pattern breakdown
        }
    """
    if not patterns:
        return {
            'signal': 0,
            'confidence': 0.0,
            'pattern_count': 0,
            'details': []
        }
    
    # Filter by minimum confidence
    strong_patterns = [p for p in patterns if p['confidence'] >= min_confidence]
    
    if not strong_patterns:
        return {
            'signal': 0,
            'confidence': 0.0,
            'pattern_count': 0,
            'details': []
        }
    
    # Calculate weighted signal
    total_weight = 0
    weighted_signal = 0
    
    for pattern in strong_patterns:
        weight = pattern['confidence']
        weighted_signal += pattern['direction'] * weight
        total_weight += weight
    
    # Normalize
    if total_weight > 0:
        avg_signal = weighted_signal / total_weight
        overall_confidence = total_weight / len(strong_patterns)
    else:
        avg_signal = 0
        overall_confidence = 0
    
    # Convert to discrete signal
    if avg_signal > 0.3:
        signal = 1  # Bullish
    elif avg_signal < -0.3:
        signal = -1  # Bearish
    else:
        signal = 0  # Neutral
    
    return {
        'signal': signal,
        'confidence': min(overall_confidence, 1.0),
        'pattern_count': len(strong_patterns),
        'details': strong_patterns
    }
